File: Python/smallestRange.py
#!/usr/bin/env python3
# encoding: utf-8

"""
@Filename: smallestRange.py
@Function: 最小区间
@Link: https://leetcode-cn.com/problems/smallest-range-covering-elements-from-k-lists/
@Python Version: 3.8
@Author: Wei Li
@Date：2021-08-30
"""

from typing import List
import heapq


class Solution:
    def smallestRange(self, nums: List[List[int]]) -> List[int]:
        rangeLeft, rangeRight = -10**9, 10**9
        maxValue = max(vec[0] for vec in nums)
        priorityQueue = [(vec[0], i, 0) for i, vec in enumerate(nums)]
        heapq.heapify(priorityQueue)

        while True:
            minValue, row, idx = heapq.heappop(priorityQueue)
            if maxValue - minValue < rangeRight - rangeLeft:
                rangeLeft, rangeRight = minValue, maxValue
            if idx == len(nums[row]) - 1:
                break
            maxValue = max(maxValue, nums[row][idx + 1])
            heapq.heappush(priorityQueue, (nums[row][idx + 1], row, idx + 1))

        return [rangeLeft, rangeRight]


# -----------------------------
if __name__ == "__main__":
    nums = [[4, 10, 15, 24, 26], [0, 9, 12, 20], [5, 18, 22, 30]]

    solution = Solution()
    smallest_range = solution.smallestRange(nums)
    print(f"The solutio of this problem is {smallest_range}")

File: C++/smallestRange.cpp
/**
 * @File    : smallestRange.cpp
 * @Brief   : 最小区间
 * @Link    : https://leetcode-cn.com/problems/smallest-range-covering-elements-from-k-lists/
 * @Author  : Wei Li
 * @Date    : 2021-08-30
*/

#include <iostream>
#include <vector>
#include <queue>
#include <algorithm>

class Solution
{
public:
    std::vector<int> smallestRange(std::vector<std::vector<int>> &nums)
    {
        int rangeLeft = 0, rangeRight = INT_MAX;
        int size = nums.size();
        std::vector<int> next(size);

        auto cmp = [&](const int &u, const int &v)
        {
            return nums[u][next[u]] > nums[v][next[v]];
        };
        std::priority_queue<int, std::vector<int>, decltype(cmp)> pq(cmp);
        int minValue = 0, maxValue = INT_MIN;
        for (int i = 0; i < size; ++i)
        {
            pq.emplace(i);
            maxValue = std::max(maxValue, nums[i][0]);
        }

        while (true)
        {
            int row = pq.top();
            pq.pop();
            minValue = nums[row][next[row]];
            if (maxValue - minValue < rangeRight - rangeLeft)
            {
                rangeLeft = minValue;
                rangeRight = maxValue;
            }
            if (next[row] == nums[row].size() - 1)
            {
                break;
            }
            ++next[row];
            maxValue = std::max(maxValue, nums[row][next[row]]);
            pq.emplace(row);
        }

        return {rangeLeft, rangeRight};
    }
};

template <typename T>
void DispalyElement(const T &container)
{
    for (auto element : container)
    {
        std::cout << element << " ";
    }
    std::cout << std::endl;
}

// -------------------------------
int main(int argc, char **argv)
{
    std::vector<std::vector<int>> nums{{1, 2, 3}, {1, 2, 3}, {1, 2, 3}};

    Solution solution;
    auto smallest_range = solution.smallestRange(nums);
    DispalyElement(smallest_range);

    return 0;
}
